src/runtime/string.cpp

```diff
@@ -680,6 +680,56 @@ static PRIMFN(prim_hash_str) {
   RETURN(String::alloc(runtime.heap, std::move(hex)));
 }
 
+static PRIMTYPE(type_to_upper) {
+  return args.size() == 1 && args[0]->unify(Data::typeString) && out->unify(Data::typeString);
+}
+
+static PRIMFN(prim_to_upper) {
+  EXPECT(1);
+  STRING(str, 0);
+
+  const char *in = str->c_str();
+  ssize_t in_size = str->length;
+  std::string out;
+  while (in_size > 0) {
+    int32_t codepoint = 0;
+    ssize_t size = utf8proc_iterate(reinterpret_cast<const uint8_t *>(in), in_size, &codepoint);
+    in += size;
+    in_size -= size;
+    int32_t upper = utf8proc_toupper(codepoint);
+    char buf[4] = {0};
+    ssize_t bytes = utf8proc_encode_char(upper, reinterpret_cast<uint8_t *>(buf));
+    out.insert(out.end(), buf, buf + bytes);
+  }
+
+  RETURN(String::alloc(runtime.heap, std::move(out)));
+}
+
+static PRIMTYPE(type_to_lower) {
```

Comment:
we 100% shouldn't do this but I just realized we could probably have a `static PRIMTYPE(type_str_to_str) { ... }`  that we share between all `String => String`  prims

Reply:
Yeah I always wonder if I should do that and then never do it lol.